File: clients/python/framework/net/host.py
import socket, threading
from framework.net.protocol import encode, decode, MsgType


class Host:
    def __init__(self, port, game=None, session_name='', game_name=None, max_players=2):
        self.port = port
        self.game = game
        self.session_name = session_name or game_name or ''
        self._game_name = game_name
        self._max_players = max_players
        self.clients = {}   # player_id -> socket
        self._lock = threading.Lock()
        self._server = None
        self._running = False

    def start(self):
        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._server.bind(('', self.port))
        self._server.listen(8)
        self._server.settimeout(1.0)
        self._running = True
        threading.Thread(target=self._accept_loop, daemon=True).start()

    def serve(self):
        self.start()
        while self._running:
            threading.Event().wait(0.1)

    def stop(self):
        self._running = False
        if self._server:
            self._server.close()

    def _accept_loop(self):
        while self._running:
            try:
                conn, _ = self._server.accept()
                threading.Thread(target=self._handle, args=(conn,), daemon=True).start()
            except socket.timeout:
                continue
            except OSError:
                break

    def _handle(self, conn):
        player_id = None
        try:
            for line in conn.makefile('rb'):
                if not line.strip():
                    continue
                try:
                    msg = decode(line)
                except Exception:
                    continue
                t = msg.get('type')
                if t == MsgType.JOIN:
                    player_id = msg['from']
                    with self._lock:
                        self.clients[player_id] = conn
                    players_now = list(self.clients.keys())
                    self.broadcast({'type': MsgType.PLAYER_LIST, 'players': players_now})
                    if (self.game is None and self._game_name and
                            len(players_now) >= self._max_players):
                        from framework.lobby.session import _load_game_classes
                        classes = _load_game_classes()
                        if self._game_name in classes:
                            self.game = classes[self._game_name]()
                            self.game.start(players_now)
                            self.broadcast({'type': MsgType.GAME_START,
                                            'game': self._game_name,
                                            'players': players_now})
                elif t == MsgType.MOVE:
                    self._handle_move(msg)
                elif t == MsgType.CHAT:
                    self.broadcast(msg)
                elif t == MsgType.LEAVE:
                    break
        finally:
            if player_id:
                with self._lock:
                    self.clients.pop(player_id, None)
                self.broadcast({'type': MsgType.PLAYER_LIST,
                                'players': list(self.clients.keys())})
            try:
                conn.close()
            except Exception:
                pass

    def _handle_move(self, msg):
        player_id = msg['from']
        move_data = msg.get('data', {})
        if self.game.current_turn() != player_id:
            from framework.i18n import t
            self._send_to(player_id, {'type': MsgType.ERROR, 'message': t('error.not_your_turn')})
            return
        if not self.game.validate_move(player_id, move_data):
            from framework.i18n import t
            self._send_to(player_id, {'type': MsgType.ERROR, 'message': t('error.invalid_move')})
            return
        self.game.apply_move(player_id, move_data)
        with self._lock:
            pids = list(self.clients.keys())
        for pid in pids:
            state_msg = {'type': MsgType.STATE,
                         'data': self.game.get_state(perspective=pid)}
            self._send_to(pid, state_msg)
        done, winner = self.game.is_over()
        if done:
            self.broadcast({'type': MsgType.GAME_OVER, 'winner': winner})

    def broadcast(self, msg):
        data = encode(msg)
        with self._lock:
            conns = list(self.clients.values())
        for c in conns:
            self._send_raw(c, data)

    def _send_to(self, player_id, msg):
        with self._lock:
            conn = self.clients.get(player_id)
        if conn:
            self._send_raw(conn, encode(msg))

    def _send_raw(self, conn, data):
        try:
            conn.sendall(data)
        except Exception:
            pass

File: clients/bun/src/framework/net/host.ts
import * as net from 'net';
import { encode, decode, MsgType, WireMsg } from './protocol';
import { loadGameClasses } from '../lobby/session';
import { t } from '../i18n';
import type { BaseGame } from '../../games/base';

export class Host {
  readonly port: number;
  private _gameName: string | null;
  private _maxPlayers: number;
  private _clients: Map<string, net.Socket>;
  private _server: net.Server | null;
  private _running: boolean;
  game: BaseGame | null;

  constructor({ port, gameName = null, maxPlayers = 2 }: { port: number; gameName?: string | null; maxPlayers?: number }) {
    this.port = port;
    this._gameName = gameName;
    this._maxPlayers = maxPlayers;
    this._clients = new Map();
    this._server = null;
    this._running = false;
    this.game = null;
  }

  start(): Promise<void> {
    this._server = net.createServer(sock => this._handleClient(sock));
    this._running = true;
    return new Promise((resolve, reject) => {
      this._server!.once('error', reject);
      this._server!.once('listening', () => {
        this._server!.removeListener('error', reject);
        resolve();
      });
      this._server!.listen(this.port);
    });
  }

  stop(): void {
    this._running = false;
    if (this._server) this._server.close();
    for (const sock of this._clients.values()) {
      try { sock.destroy(); } catch (_) {}
    }
  }

  private _handleClient(sock: net.Socket): void {
    let buf = '';
    let playerId: string | null = null;

    sock.on('data', (chunk: Buffer) => {
      buf += chunk.toString('utf8');
      const lines = buf.split('\n');
      buf = lines.pop() ?? '';
      for (const line of lines) {
        if (!line.trim()) continue;
        let msg: WireMsg;
        try { msg = decode(line); } catch (_) { continue; }
        this._handleMessage(sock, msg, (pid: string) => { playerId = pid; });
      }
    });

    sock.on('close', () => {
      if (playerId) {
        this._clients.delete(playerId);
        this.broadcast({ type: MsgType.PLAYER_LIST, players: [...this._clients.keys()] });
      }
    });

    sock.on('error', () => {});
  }

  private _handleMessage(sock: net.Socket, msg: WireMsg, setId: (pid: string) => void): void {
    const type = msg.type;
    if (type === MsgType.JOIN) {
      const pid = msg.from!;
      setId(pid);
      this._clients.set(pid, sock);
      const players = [...this._clients.keys()];
      this.broadcast({ type: MsgType.PLAYER_LIST, players });
      if (!this.game && this._gameName && players.length >= this._maxPlayers) {
        try {
          const classes = loadGameClasses();
          if (classes[this._gameName]) {
            this.game = new classes[this._gameName]();
            this.game.start(players);
            this.broadcast({ type: MsgType.GAME_START, game: this._gameName, players });
          }
        } catch (e: unknown) {
          process.stderr.write(`[host] game start error: ${(e as Error).message}\n`);
        }
      }
    } else if (type === MsgType.MOVE) {
      this._handleMove(msg);
    } else if (type === MsgType.CHAT) {
      this.broadcast(msg);
    } else if (type === MsgType.LEAVE) {
      sock.destroy();
    }
  }

  private _handleMove(msg: WireMsg): void {
    const pid = msg.from!;
    if (!this.game) return;
    if (this.game.currentTurn() !== pid) {
      this._sendTo(pid, { type: MsgType.ERROR, message: t('error.not_your_turn') });
      return;
    }
    const moveData = (msg.data ?? {}) as Record<string, unknown>;
    if (!this.game.validateMove(pid, moveData)) {
      this._sendTo(pid, { type: MsgType.ERROR, message: t('error.invalid_move') });
      return;
    }
    this.game.applyMove(pid, moveData);
    for (const p of this._clients.keys()) {
      this._sendTo(p, { type: MsgType.STATE, data: this.game.getState(p) as Record<string, unknown> });
    }
    const [done, winner] = this.game.isOver();
    if (done) this.broadcast({ type: MsgType.GAME_OVER, winner });
  }

  broadcast(msg: WireMsg): void {
    const data = encode(msg);
    for (const sock of this._clients.values()) {
      try { sock.write(data); } catch (_) {}
    }
  }

  private _sendTo(playerId: string, msg: WireMsg): void {
    const sock = this._clients.get(playerId);
    if (sock) {
      try { sock.write(encode(msg)); } catch (_) {}
    }
  }
}
